rpc/http.go

```diff
@@ -313,3 +97,94 @@ func (client *httpClient) Close() {
 func (client *httpClient) SupportedModules() (map[string]string, error) {
 	return SupportedModules(client)
 }
+
+// httpStream reads and writes HTTP requests and responses and can be used
+// as the ReadWriteCloser for the ServerCodec.
+type httpStream struct {
+	mu              sync.Mutex
+	request         *http.Request
+	firstRead     bool
+	response        http.ResponseWriter
+	responseWritten chan struct{}
+}
+
+// Read the body of the next HTTP request.
+func (rw *httpStream) Read(p []byte) (int, error) {
+	n, err := rw.request.Body.Read(p)
+	if n == 0 && err == io.EOF && !rw.firstRead { // request already read
+		<-rw.responseWritten // wait for response to be written (http is sync)
+		return 0, io.EOF     // indicate that the request has been read
+	}
+	rw.firstRead = false
+	return n, err
+}
+
+// Write the given buffer as a HTTP response.
+// Note: this will write the headers which won't allow the response to be
+// written in multiple chunks.
+func (rw *httpStream) Write(p []byte) (int, error) {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
+	if rw.responseWritten != nil {
+		n, err := rw.response.Write(p)
+		close(rw.responseWritten)
+		rw.responseWritten = nil
+		return n, err
+	}
+	return 0, errors.New("response already written")
+}
+
+// Close indicates that the HTTP response has been sent.
+func (rw *httpStream) Close() error {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
+	if rw.responseWritten != nil {
+		close(rw.responseWritten)
+		rw.responseWritten = nil
+	}
+	return nil
+}
+
+// newJSONHTTPHandler creates a HTTP handler that will parse incoming JSON requests,
+// send the request to the given API provider and sends the response back to the caller.
+func newJSONHTTPHandler(srv *Server) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.ContentLength > maxHTTPRequestContentLength {
+			http.Error(w,
+				fmt.Sprintf("content length too large (%d>%d)", r.ContentLength, maxHTTPRequestContentLength),
+				http.StatusRequestEntityTooLarge)
+			return
+		}
+
+		// turn message based HTTP into a stream that can be used by the ServerCodec
+		rw := &httpStream{
+			request:         r,
+			response:        w,
+			firstRead:       true,
+			responseWritten: make(chan struct{}),
+		}
+
+		codec := NewJSONCodec(rw)
+		w.Header().Set("content-type", "application/json")
+
+		go srv.ServeCodec(codec)
+
+		<-rw.responseWritten
+
+		return
+	}
+}
+
+// NewHTTPServer creates a new HTTP RPC server around an API provider.
+func NewHTTPServer(corsString string, srv *Server) *http.Server {
+	c := cors.New(cors.Options{
+		AllowedOrigins: strings.Split(corsString, " "),
```

Comment:
This should be comma separated, not? We've converted all your CLI flags to comma separated versions since spaces/sorrounding-quotes lead to weird issues when transitioning the flags through various environments (e.g. bash, docker). The previous code also had comma here.